workflow/controller/cache.go

```diff
@@ -0,0 +1,117 @@
+package controller
+
+import (
+	"encoding/json"
+	"regexp"
+
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+type MemoizationCache interface {
+	Load(key string) (*wfv1.Outputs, bool)
+	Save(key string, node_id string, value *wfv1.Outputs) bool
+}
+
+type CacheEntry struct {
+	ExpiresAt string       `json:"expiresAt"`
+	NodeID    string       `json:"nodeID"`
+	Outputs   wfv1.Outputs `json:"outputs"`
+}
+
+type configMapCache struct {
+	namespace     string
+	configMapName string
+	kubeClient    kubernetes.Interface
+}
+
+func NewConfigMapCache(cm string, ns string, ki kubernetes.Interface) MemoizationCache {
+	return &configMapCache{
+		configMapName: cm,
+		namespace:     ns,
+		kubeClient:    ki,
+	}
+}
+
+func generateCacheKey(key string) string {
+	log.Infof("Validating cache key %s", key)
+	reg, err := regexp.Compile("[^-._a-zA-Z0-9]+")
+	if err != nil {
+		log.Fatal(err)
```

Comment:
this will panic the controller - you should return an error instead

Reply:
This is not an issue, since the expression of the regex is constant. If it compiles once (as in the dev environment) it will always compile. More over, to avoid having to deal with the error, you can use `regex.MustCompile` instead as we do elsewhere in the codebase.